File: javascript/binary-search-tree.js
class BinaryTree {
  constructor(initialValue) {
    this.child = { left: null, right: null };
    if (initialValue) {
      this.value = initialValue;
    }
  }

  insert(item) {
    if (!this.value) {
      this.value = item;
      return;
    }

    this.createNode(
      this.compareNodes(this.value, item) ? "left" : "right",
      item
    );
  }

  search(key) {}

  createNode(target, item) {
    if (!this.child[target]) {
      this.child[target] = new BinaryTree(item);
    } else {
      this.child[target].insert(item);
    }
  }

  compareNodes(a, b) {
    return a > b;
  }

  printPre() {
    process.stdout.write(` ${this.value}`);
    this.child.left && this.child.left.printPre();
    this.child.right && this.child.right.printPre();
  }

  printIn() {
    this.child.left && this.child.left.printIn();
    process.stdout.write(` ${this.value}`);
    this.child.right && this.child.right.printIn();
  }

  printPost() {
    this.child.left && this.child.left.printPost();
    this.child.right && this.child.right.printPost();
    process.stdout.write(` ${this.value}`);
  }
}

function processEntry(input, index) {
  const tree = new BinaryTree();

  input.split(" ").forEach((item) => {
    tree.insert(parseInt(item));
  });

  process.stdout.write(`Case ${index + 1}:\n`);

  process.stdout.write("Pre.:");
  tree.printPre();
  process.stdout.write("\n");

  process.stdout.write("In..:");
  tree.printIn();
  process.stdout.write("\n");

  process.stdout.write("Post:");
  tree.printPost();
  process.stdout.write("\n");
}

main();
function main() {
  const input = require("fs").readFileSync("/dev/stdin", "utf8");
  const lines = input.split("\n");

  const numIterations = lines.shift();

  for (let i = 0; i < numIterations; i++) {
    const _ = lines.shift();
    const line = lines.shift();
    processEntry(line, i);
    console.log("");
  }
}

File: cpp/binary-search-tree.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

template <typename T>
class BinaryTree {
  public:
  BinaryTree *left;
  BinaryTree *right;
  bool initialized;
  T value;

  BinaryTree() {
    this->initialized = false;
    this->left = NULL;
    this->right = NULL;
  }

  BinaryTree(vector<T> values) {
    this->left = NULL;
    this->right = NULL;
    this->initialized = true;
    this->insert(values);
  }

  BinaryTree(T value) {
    this->left = NULL;
    this->right = NULL;
    this->initialized = true;
    this->value = value;
  }

  bool insert(T value) {
    if (!this->initialized) {
      this->value = value;
      this->initialized = true;
      return true;
    }

    this->createNode(
      this->compareNodes(value, this->value) ? "right" : "left", 
      value
    );
    return true;
  }

  bool insert(vector<T> values) {
    for (T item : values) {
      if (!this->insert(item)) {
        return false;
      }
    }

    return true;
  }

  BinaryTree** getNode(string target) {
    if (target == "right") {
      return &this->right;
    }
    return &this->left;
  }

  bool createNode(string target, T item) {
    BinaryTree** targetNode = this->getNode(target); 
    if (!*targetNode) {
      *targetNode = new BinaryTree(item);
    } else {
      if (!(*targetNode)->insert(item)) {
        return false;
      }
    }

    return true;
  }

  bool compareNodes(T a, T b) {
    return a > b;
  }

  void printPre() {
    cout << " " << this->value;
    if (this->left) this->left->printPre();
    if (this->right) this->right->printPre();
  }

  void printIn() {
    if (this->left) this->left->printIn();
    cout << " " << this->value;
    if (this->right) this->right->printIn();
  }

  void printPost() {
    if (this->left) this->left->printPost();
    if (this->right) this->right->printPost();
    cout << " " << this->value;
  }
};

void processEntry(vector<int> entry, int index) {
  BinaryTree<int> bt;
  bt.insert(entry);

  cout << "Case " << index << ":" << endl;
  cout << "Pre.:";
  bt.printPre();
  cout << endl;
  cout << "In..:";
  bt.printIn();
  cout << endl;
  cout << "Post:";
  bt.printPost();
  cout << endl;
  cout << endl;
}

int main() {
  int numTests, size;  

  cin >> numTests;
  for(int i = 1; i <= numTests; i++) {
    cin >> size;

    vector<int> entry; 
    while(size--) {
      int num;
      cin >> num;

      entry.push_back(num);
    }

    processEntry(entry, i);
  }
}